### PR TITLE
chore: upgrade @aws-sdk/client-sqs to version which supports sqs json protocol

### PR DESCRIPTION
**Pull request by petrisorciprian-vitals**
**Description:**
Updates the aws sdk to a version which supports the [AWS SQS JSON protocol](https://aws.amazon.com/about-aws/whats-new/2023/11/amazon-sqs-support-json-protocol/).

**Type of change:**

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

**Why is this change required?:**
It enables better performance due to usage of the AWS SQS JSON communication protocol instead of the AWS SQS Query protocol between client and server.

**Code changes:**

- update of dependencies
---

**Checklist:**

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by petrisorciprian-vitals** [comment]
The localstack upgrade is required since the previous version did not support the SQS JSON protocol as it was old.
The healthcheck path and init script path also changed with in more recent versions of localstack and it is unnecessary to mount a tmp dir.

More recent versions of docker also have do

**Comment by nicholasgriffintn** [comment]
Our CLA check is broken for some reason so I'll just have to do that manually...

> Thank you for your submission, we really appreciate it. Like many open-source projects, we ask that you sign our [Contributor License Agreement](https://bbc.github.io/cla-signatures/cla/v1/cla.html) before we can accept your contribution. You can sign the CLA by just posting a Pull Request Comment reply with the following:""

>"I have read the CLA Document and I hereby sign the CLA"

**Comment by dciprian-petrisor** [comment]
I have read the CLA Document and I hereby sign the CLA

File: src/consumer.ts [hunk 1]
```diff
@@ -198,7 +199,7 @@ export class Consumer extends TypedEventEmitter {
     let currentPollingTimeout = this.pollingWaitTimeMs;
     this.receiveMessage({
       QueueUrl: this.queueUrl,
-      AttributeNames: this.attributeNames,
+      AttributeNames: this.attributeNames as QueueAttributeName[],
```
Comment (nicholasgriffintn) [inline src/consumer.ts:202]: What's the reason for this, is there a new TS error here? One that I might have to look into as there might be a better way of typing this.

File: test/scripts/docker-compose.yml [hunk 2]
```diff
@@ -2,8 +2,7 @@ version: '3'
 services:
   localstack:
     container_name: local_sqs_aws
-    image: localstack/localstack:1.2.0@sha256:41216e34e25546654c58b8075b133027eab3d958e605ef2e7af168d1f05a9364
-    command: sh -c "/usr/local/bin/init.sh && exec myexecutable"
+    image: localstack/localstack:latest@sha256:da35c19b83b646b2f4511b62f39689bc73cf0fe53ac4159dfd513e9774356324
```
Comment (nicholasgriffintn) [inline test/scripts/docker-compose.yml:5]: I believe localstack will soon release a fixed version of their Docker image, so we may be better off waiting for that, I'd prefer not to use latest.

File: test/scripts/docker-compose.yml [hunk 3]
```diff
@@ -17,11 +16,10 @@ services:
     ports:
       - "4566-4599:4566-4599"
     volumes:
-      - '${TMPDIR:-/tmp/localstack}:/tmp/localstack'
       - '/var/run/docker.sock:/var/run/docker.sock'
-      - ./localstack:/docker-entrypoint-initaws.d
+      - ./localstack:/etc/localstack/init/ready.d
     healthcheck:
-      test: curl http://localhost:4566/health || exit 1
+      test: curl http://localhost:4566/_localstack/health || exit 1
```
Comment (nicholasgriffintn) [inline test/scripts/docker-compose.yml:22]: I'll need to read up on these changes and the removal of `command`

**Comment by nicholasgriffintn** [comment]
To note: This would require a major version of the package as the new version of aws-sdk is breaking against anything that's expects the XML protocol.

File: src/consumer.ts (hunk 1 above)
Comment (petrisorciprian-vitals) [reply src/consumer.ts:202]: Yes, there's a new typing added for that field, so the existing this.attributeNames (which is string[]  typed) was raising an error there.

I wasn't sure if we should change the typing for this.attributeNames or not to match; I leave this decision up to you.

File: test/scripts/docker-compose.yml (hunk 3 above)
Comment (petrisorciprian-vitals) [reply test/scripts/docker-compose.yml:22]: Not sure what that command block was doing anyway; perhaps you can explain a bit?
Localstack launches as expected and runs the init.sh hook mounted in /etc/localstack/init/ready.d; is there anything else that needs to happen before running the integration tests?

Comment (nicholasgriffintn) [reply test/scripts/docker-compose.yml:22]: Honestly I can't remember 😆

File: test/scripts/docker-compose.yml (hunk 2 above)
Comment (petrisorciprian-vitals) [reply test/scripts/docker-compose.yml:5]: As we were discussing this, they just released v3 of localstack alongside a 'stable' and '3' tag.
I can update the MR to use localstack/localstack:stable or localstack/localstack:3 instead.
Let me know if you want me to do so.

Comment (petrisorciprian-vitals) [reply test/scripts/docker-compose.yml:5]: I've just checked with both 'stable' and '3' tagged releases of localstack, integration tests pass; I await your decision for this.

**Comment by petrisorciprian-vitals** [comment]
@nicholasgriffintn 
I've pushed a new commit which removes some obsolete envs from the localstack service in docker-compose.yml
See https://docs.localstack.cloud/references/configuration/#legacy for details.

File: test/scripts/docker-compose.yml (hunk 3 above)
Comment (petrisorciprian-vitals) [reply test/scripts/docker-compose.yml]: I've updated the healthcheck CMD to be a bit nicer; it checks if SQS service specifically is in 'running' now.

**Comment by nicholasgriffintn** [comment]
Hey thanks for the updates and sorry, I've been super busy this week, I'm gonna slot some time to look into this on Friday hopefully, if not before.

File: test/scripts/docker-compose.yml (hunk 2 above)
Comment (nicholasgriffintn) [reply test/scripts/docker-compose.yml]: Committed a change to fix the version just now.

File: test/scripts/docker-compose.yml (hunk 3 above)
Comment (nicholasgriffintn) [reply test/scripts/docker-compose.yml]: Seems to still spin up the docker, so this appears to be fine.

File: src/consumer.ts (hunk 1 above)
Comment (nicholasgriffintn) [reply src/consumer.ts]: I've pushed a commit up to change or types to respect this change.

**Comment by nicholasgriffintn** [comment]
> **CLA Assistant Lite bot:** Thank you for your submission, we really appreciate it. Like many open-source projects, we ask that you all sign our [Contributor License Agreement](https://bbc.github.io/cla-signatures/cla/v1/cla.html) before we can accept your contribution. You can sign the CLA by just posting a Pull Request Comment same as the below format.
> 
> I have read the CLA Document and I hereby sign the CLA
> 
> **1** out of **2** committers have signed the CLA.✅ (nicholasgriffintn)[https://github.com/nicholasgriffintn]❌ @petrisorciprian-vitalsYou can retrigger this bot by commenting **recheck** in this Pull Request

This was signed here: https://github.com/bbc/sqs-consumer/pull/440#issuecomment-1814101497

**Review by nicholasgriffintn — APPROVED** [review]
It's shipping time

**Pull request merged**